e2e-tests/src/validators.rs

```diff
@@ -138,9 +138,20 @@ pub async fn prepare_validators<S: SignedConnectionApi + AuthorRpc>(
         }));
     }
 
-    for controller in accounts.controller_raw_keys.iter() {
+    let address_tokens = node.split(':').collect::<Vec<_>>();
+    let ws = address_tokens[0];
+    let ip_address = address_tokens[1];
+    let first_port = address_tokens[2].parse::<usize>()?;
+
+    // Assumes the same ip address and consecutive ports for nodes, e.g. ws://127.0.0.1:9943,
+    // ws://127.0.0.1:9944, etc.
```

Comment:
This part seems out of place in this function, shouldn't we pass in a set of connections instead, or at least a set of addresses?

Reply:
Sure, I have extracted this to a separate function and passed the connections in.